overlord/devicestate/devicemgr.go

```diff
@@ -329,22 +335,74 @@ func (m *DeviceManager) maybeSetupUbuntuSave() error {
 	if !runMntSaveMounted {
 		// we don't have ubuntu-save, save will be used directly
 		logger.Noticef("no ubuntu-save mount")
-		m.saveAvailable = true
 		return nil
 	}
 
-	logger.Noticef("bind-mounting ubuntu-save under %v", dirs.SnapSaveDir)
+	// In newer core20/core22 we have a mount unit for ubuntu-save, which we
+	// we will try to start first. Invoking systemd-mount in this case would fail.
+	err = exec.Command("systemctl", "enable", "var-lib-snapd-save.mount", "--now").Run()
+	if err == nil {
+		logger.Noticef("mount unit for ubuntu-save was started")
+		return nil
+	}
 
+	// Otherwise try to directly mount the partition with systemd-mount.
+	logger.Noticef("bind-mounting ubuntu-save under %v", dirs.SnapSaveDir)
 	err = systemd.New(systemd.SystemMode, progress.Null).Mount(boot.InitramfsUbuntuSaveDir,
 		dirs.SnapSaveDir, "-o", "bind")
 	if err != nil {
 		logger.Noticef("bind-mounting ubuntu-save failed %v", err)
 		return fmt.Errorf("cannot bind mount %v under %v: %v", boot.InitramfsUbuntuSaveDir, dirs.SnapSaveDir, err)
 	}
-	m.saveAvailable = true
 	return nil
 }
 
+// ensureUbuntuSaveSnapFolders creates the necessary folder structure for
+// /var/lib/snapd/save/snap/<snap>. This is normally done during installation
+// of a snap, but there are two cases where this can be insufficient.
+//
+// 1. When migrating to a newer snapd, folders are not automatically created for
+//    snaps that are already installed. They will be created during a refresh, but
+//    we should not rely on that uncertainty.
```

Comment:
If I understand correctly, I'd reformulate the comment like this:
```suggestion
//    snaps that are already installed. They will only be created during a refresh of
//    the snap itself, whereas we want to cover all the cases.
```

Reply:
Thank you, always really appreciate input on this